### ISSUE TITLE
Output changes if different sample IDs are assigned

### ISSUE THREAD

**Issue by ryuma-matsubara** [simonlabcode/bam2bakR#15]
Suppose there are 8 bam files. In the example below, bam files with 'WO' are from RNA-seq without 4sU treatment.  
I analyzed the same data (same bam files) with two different sets of sample IDs:

Example yaml 1 
```
samples:
  s_1: ../../bam2bakr_working/7.bam
  s_2: ../../bam2bakr_working/8.bam
  s_3: ../../bam2bakr_working/9.bam
  s_4: ../../bam2bakr_working/10.bam
  ctr_5: ../../bam2bakr_working/7WO.bam
  ctr_6: ../../bam2bakr_working/8WO.bam
  ctr_7:  ../../bam2bakr_working/9WO.bam
  ctr_8: ../../bam2bakr_working/10WO.bam

control_samples: ['ctr_7', 'ctr_8']
```

Example yaml 2
```
samples:
  s300_1: ../../bam2bakr_working/7.bam
  s300_2: ../../bam2bakr_working/8.bam
  s400_1: ../../bam2bakr_working/9.bam
  s400_2: ../../bam2bakr_working/10.bam
  noTUC_s300_1: ../../bam2bakr_working/7WO.bam
  noTUC_s300_2: ../../bam2bakr_working/8WO.bam
  noTUC_s400_1:  ../../bam2bakr_working/9WO.bam
  noTUC_s400_2: ../../bam2bakr_working/10WO.bam

control_samples: ['noTUC_s400_1', 'noTUC_s400_2']
```

Following Bam2bakR instruction, I counted per-base mutation rate:  

```
cB <- fread('cB.csv.gz')
TUCrate <- cB[ ,.(mutation_rate = sum(TC*n)/sum(nT*n)), by = sample]
ggplot(TUCrate, aes(x=sample, y=mutation_rate)) + geom_col()
```

and surprisingly found that the outputs are quite different (sorry for ugly plots).  
Example yaml 1 | Example yaml 2
--- | ---
![example1](https://github.com/simonlabcode/bam2bakR/assets/101512284/b8ac15b2-0e54-427c-9d44-ad9163bb2831=20%x)  | ![example2](https://github.com/simonlabcode/bam2bakR/assets/101512284/f225cbe6-ffa8-4353-9934-4743c829e121=20%x)

Additionally, the example yaml 2 produces an output in which none of 's400_2' was assigned to any features:
```
cB |> filter(sample %in% 's400_2') |> select(GF) |> is.na() |> all()
# [1] TRUE
```

I assume there is a hidden rule for how (not) to name samples? It is, in my perspective, scary as no error message was produced. (it certainly gave me a heart attack) I'd appreciate it if you would take a look at this problem.  

Best regards, 
Ryuma

**Comment by isaacvock** [comment]
Hello Ryuma,

This is strange, as there should be no hidden sample naming rules. Changes have been made to the pipeline in the last several days; were these pipeline outputs produced many days apart? 

More information will be necessary if I am to identify the problem. Can you please provide:

1) The full config files
2) logs/featurecounts_genes/s400_2.log from the second run where no reads were assigned to features. The equivalent file from the first run would also be appreciated 
3) logs/merge_features_and_muts/s400_2.log and it's equivalent file from the first run
4) if you could downsample the s400_2 bam file and send me that downsampled file, that would be very helpful in debugging.

If you have time, please check all of the .log files for any error messages. It is possible for errors to crop up when executing some of the shell scripts used by the pipeline but for the pipeline not to crash.

In the meantime, I will try to reproduce the problem with my own set of bam files.

Isaac

**Comment by isaacvock** [comment]
As an update, I was able to reproduce the problem with my own bam files using the same naming convention as you. I am looking into what is going wrong now.

**Comment by isaacvock** [comment]
I figured out the problem and have pushed a fix to the main branch. 

The problem was in the merge_features_and_muts steps where I grep for sample IDs in .featureCounts CORE files. Given the regex pattern I was using, if one sample ID was a subset of the end of another (e.g., s400_1 is a subset of the end of noTUC_s400_1) then this could lead to the wrong files getting loaded. Because of this, the output that you get with the first config is the correct output. I have changed the regex pattern to avoid this problem. 

Thank you for bringing this to my attention and let me know if the fix addresses the problem for you,
Isaac
